brave/src/main/java/brave/secondary_sampling/SecondarySampler.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.secondary_sampling;
+
+import brave.http.HttpServerRequest;
+import brave.propagation.TraceContext;
+
+public interface SecondarySampler {
+  /**
+   * Returns a trigger function that decides if this request is sampled for the given {@code
+   * samplingKey}.
+   *
+   * <p>This is invoked during {@link TraceContext.Extractor#extract(Object)}. It is usually the
+   * case that the carrier parameter is a request. For example, in http middleware, it will be
+   * {@link HttpServerRequest}. However, in other RPC middleware it may be a different type.
+   *
+   * <p>For example, to only handle http requests, implement like so:
+   * <pre>{@code
+   *     if (!(carrier instanceof HttpServerRequest)) return false;
+   *     HttpServerRequest request = (HttpServerRequest) carrier;
+   *     return isSampled(request, builder.samplingKey());
+   * }</pre>
+   *
+   * <p><h3>Mutating the builder</h3>
+   * A side-effect of a sampling decision could be adding parameters or refreshing a {@link
+   * SecondarySamplingState.Builder#ttl(int)}.
+   *
+   * @param request the incoming request type, such as {@link HttpServerRequest}
+   * @param builder state extracted for this sampling key.
+   * @return true if the {@link SecondarySamplingState.Builder#samplingKey()} is sampled for this
```

Comment:
Same as the previous [comment](https://github.com/openzipkin-contrib/zipkin-secondary-sampling/pull/7/files#r328250029).

The presence of the sampling key means that the request is already sampled by upstream. Now this service can participate or not by matching sampling_key with a local dynamic config or check the `ttl` value.

If you think the implementation should allow starting a trace by examining the request object, then the builder should allow creating a new `sampling_key`.

Reply:
It doesnt' mean it was sampled upstream, all it means is that there was a key created. The word+tense "sampled" in brave means literally this span is being recorded. Please read my comment above.

Also please bear in mind that this pull request was not created to add a new feature of creating a new sampling key. That's a separate api and would need tests to explain its usage. It is generally the culture to perform chunks of work separately as that eases progress. For example, features should not be tacked on as comments as that can overwhelm the work actually created. IF you think of something not done, and there's no  issue for it, for example provisioning a local sampling key, please raise a separate issue for that. The scope of this pull request is not to complete all work, rather just to refactor the internals such that future work can be done.